go/vt/vtgate/gateway.go

```diff
@@ -0,0 +1,78 @@
+	"flag"
+	"time"
+
+	log "github.com/golang/glog"
+	"golang.org/x/net/context"
+
+	mproto "github.com/youtube/vitess/go/mysql/proto"
+	"github.com/youtube/vitess/go/stats"
+	pb "github.com/youtube/vitess/go/vt/proto/topodata"
+	tproto "github.com/youtube/vitess/go/vt/tabletserver/proto"
+	"github.com/youtube/vitess/go/vt/tabletserver/tabletconn"
+)
+
+var (
+	// GatewayImplementation controls the implementation of Gateway.
+	GatewayImplementation = flag.String("gateway_implementation", "shardconn", "The implementation of gateway")
+)
+
+// A Gateway is the query processing module for each shard,
+// which is used by ScatterConn.
+type Gateway interface {
+	// Dial creates the connection for the specified keyspace, shard, and tablet type.
+	Dial(ctx context.Context, keyspace, shard string, tabletType pb.TabletType) error
+
+	// Execute executes the non-streaming query for the specified keyspace, shard, and tablet type.
+	Execute(ctx context.Context, keyspace, shard string, tabletType pb.TabletType, query string, bindVars map[string]interface{}, transactionID int64) (*mproto.QueryResult, error)
+
+	// ExecuteBatch executes a group of queries for the specified keyspace, shard, and tablet type.
+	ExecuteBatch(ctx context.Context, keyspace, shard string, tabletType pb.TabletType, queries []tproto.BoundQuery, asTransaction bool, transactionID int64) (*tproto.QueryResultList, error)
+
+	// StreamExecute executes a streaming query for the specified keyspace, shard, and tablet type.
+	StreamExecute(ctx context.Context, keyspace, shard string, tabletType pb.TabletType, query string, bindVars map[string]interface{}, transactionID int64) (<-chan *mproto.QueryResult, tabletconn.ErrFunc)
+
+	// Begin starts a transaction for the specified keyspace, shard, and tablet type.
+	// It returns the transaction ID.
+	Begin(ctx context.Context, keyspace string, shard string, tabletType pb.TabletType) (int64, error)
+
+	// Commit commits the current transaction for the specified keyspace, shard, and tablet type.
+	Commit(ctx context.Context, keyspace, shard string, tabletType pb.TabletType, transactionID int64) error
+
+	// Rollback rolls back the current transaction for the specified keyspace, shard, and tablet type.
+	Rollback(ctx context.Context, keyspace, shard string, tabletType pb.TabletType, transactionID int64) error
+
+	// SplitQuery splits a query into sub-queries for the specified keyspace, shard, and tablet type.
+	SplitQuery(ctx context.Context, keyspace, shard string, tabletType pb.TabletType, sql string, bindVariables map[string]interface{}, splitColumn string, splitCount int) ([]tproto.QuerySplit, error)
+
+	// Close shuts down underlying connections.
+	Close() error
+}
+
+// GatewayCreator is the func which can create the actual gateway object.
+type GatewayCreator func(serv SrvTopoServer, cell string, retryDelay time.Duration, retryCount int, connTimeoutTotal, connTimeoutPerConn, connLife time.Duration, connTimings *stats.MultiTimings) Gateway
```

Comment:
I'm not sure all these parameters will make sense in all implementations, so let's keep an eye for this and be open to move them around lower (like using flags that are specific to one implementation).